File: NC/merge-k-sorted-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution
{
public:
    struct comp
    {
        bool operator()(ListNode *a, ListNode *b)
        {
            return a->val > b->val;
        }
    };

    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        priority_queue<ListNode *, vector<ListNode *>, comp> pq;

        for (auto i : lists)
        {
            pq.push(i);
        }

        ListNode *head = NULL, *curr = NULL, *prev;

        while (pq.size() > 0)
        {
            if (head == NULL)
            {
                head = pq.top();
                curr = pq.top();
            }
            else
            {
                prev = curr;
                curr = pq.top();
                prev->next = curr;
            }
            pq.pop();
            if (curr->next)
            {
                pq.push(curr->next);
            }
        }

        return head;
    }
};

File: NC/java/MergeKSortedLinkedLists.java
package NC.java;

public class MergeKSortedLinkedLists {
    public ListNode mergeKLists(ListNode[] lists) {

        if (lists.length == 0) {
            return null;
        }

        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> {
            return a.val - b.val;
        });

        for (var node : lists) {
            pq.add(node);
        }

        ListNode ans = pq.poll();
        ListNode curr = ans;

        if (curr.next != null) {
            pq.add(curr.next);
        }

        while (pq.size() > 1) {
            ListNode temp = pq.poll();
            curr.next = temp;
            curr = temp;
            if (temp.next != null) {
                pq.add(temp.next);
            }
        }

        if (pq.size() == 1) {
            curr.next = pq.poll();
        }

        return ans;
    }
}
